src/rgw/rgw_orphan.cc

```diff
@@ -1332,32 +1333,54 @@ int RGWRadosList::run(const DoutPrefixProvider *dpp)
     return ret;
   }
 
-  const int max_keys = 1000;
+  constexpr int max_keys = 1000;
   bool truncated = true;
+  bool warned_indexless = false;
 
   do {
     std::list<std::string> buckets;
     ret = store->meta_list_keys_next(dpp, handle, max_keys, buckets, &truncated);
 
     for (std::string& bucket_id : buckets) {
-      ret = run(dpp, bucket_id);
+      ret = run(dpp, bucket_id, true);
       if (ret == -ENOENT) {
 	continue;
+      } else if (ret == -EINVAL) {
+	if (! warned_indexless) {
+	  if (yes_i_really_mean_it) {
+	    std::cerr <<
+	      "WARNING: because there is at least one indexless bucket (" <<
+	      bucket_id <<
+	      ") the results of radoslist are *incomplete*; continuing due to --yes-i-really-mean-it" <<
+	      std::endl;
+	    warned_indexless = true;
+	  } else {
+	    std::cerr << "ERROR: because there is at least one indexless bucket (" <<
+	      bucket_id <<
+	      ") the results of radoslist are *incomplete*; use --yes-i-really-mean-it to bypass error" <<
+	      std::endl;
+	    return ret;
+	  }
+	}
+	continue;
       } else if (ret < 0) {
 	return ret;
       }
     }
   } while (truncated);
 
   return 0;
-} // RGWRadosList::run()
+} // RGWRadosList::run(bool)
 
 
-int RGWRadosList::run(const DoutPrefixProvider *dpp, const std::string& start_bucket_name)
+int RGWRadosList::run(const DoutPrefixProvider *dpp,
+		      const std::string& start_bucket_name,
+		      const bool silent_indexless)
 {
   RGWObjectCtx obj_ctx(store);
   std::unique_ptr<rgw::sal::Bucket> bucket;
   int ret;
+  RGWBucketInfo bucket_info; // needed after while loop, but declared here due to goto
```

Comment:
the single `goto done` in this function looks pretty useless, maybe it's worth getting rid of it? up to you
```c++
done:
  return 0;
```
regardless, it doesn't look like `bucket_info` is ever initialized in this function. it's accessed inside the while loop with `bucket_info.is_indexless()`, but given this:
```c++
    std::unique_ptr<rgw::sal::Bucket> bucket;
    ret = store->get_bucket(dpp, nullptr, tenant_name, bucket_name, &bucket, null_yield);
```
we'd need to look at `bucket->get_info().is_indexless()` instead

(thanks for that comment about goto, i would have missed this otherwise)

Reply:
Good catch! That led me to clean up an unused variable.